Added tests for multiline text support

ANAMASGARD opened the pull request:
Add Multi-line Text Support Throughout animint2

Fixes #221

This PR enables newline characters (\n) to work in all text elements of animint2 visualizations. Users can now create multi-line text in plot titles, axis labels, legend titles, and geom_text labels.

**What Now Works:**
✅ Plot titles - ggtitle("Line 1\nLine 2")
✅ X-axis titles - xlab("X Axis\nSubtitle")
✅ Y-axis titles - ylab("Y Axis\nSubtitle")
✅ Legend titles - scale_color_discrete(name="Legend\nTitle")
✅ Text labels - geom_text(aes(label="Text\nLabel"))
✅ Tooltips - Already worked, still works (backward compatible)


Below is the screenshots of the scripts i ran locally to verify the test locally they fail as aspected
<img width="2976" height="1662" alt="Screenshot From 2025-10-21 19-08-38" src="https://github.com/user-attachments/assets/90b2fd06-f39a-413e-9972-081aecb92a14" />
<img width="2973" height="1654" alt="Screenshot From 2025-10-21 19-08-44" src="https://github.com/user-attachments/assets/d3b6e95b-7b44-427d-aa91-daa3ddcca390" />

ANAMASGARD commented:
Ran tests locally - all 19 multiline text tests pass successfully. The test failures shown are due to missing optional packages (mapproj, sp, XML) and GitHub API permissions, not related to this PR. Sir can you please help me figure out on how to pass the JS_Coverage test suite .

tdhock commented:
To get this PR to close the corresponding issue, please add

Closes #221 

on its own line, in the first comment of the PR.

tdhock commented:
Also in https://github.com/animint/animint2/pull/261#issuecomment-3427511333 it seems like there is an inconsistency between the X and Y axes.

* there is almost no space between the Y axis title (Vertical Position) and the Y axis ticks (2 3 4)
* there is a big space between the X axis title (X Axis) and the X axis tick (3).

can this be adjusted to be more consistent please?
also it should be consistent with the case of single line X/y axis title, as below from https://5-merge--animint-manual-en.netlify.app/ch06/#aes-tooltip which was created with current master
<img width="435" height="397" alt="image" src="https://github.com/user-attachments/assets/37ebd90b-f845-4b94-8fa0-b1559b83dd3b" />

tdhock reviewed (changes requested):
please add tests then fix

tdhock reviewed (changes requested):
please simplify height/width calculation

ANAMASGARD commented on inst/htmljs/animint.js:
```diff
@@ -127,20 +127,79 @@ var animint = function (to_select, json_file) {
+          .attr('dy', i === 0 ? 0 : lineHeight + 'em')
+          .text(line);
+      });
+      
```
Done , please review and also give your feedback if anything else need to be changed .

tdhock commented on inst/htmljs/animint.js:
```diff
@@ -125,24 +125,71 @@ var animint = function (to_select, json_file) {
+      setMultilineText(textElement, pText);
+    } else {
+      // Single line: simple text rendering
+      textElement.text(pText);
```
it seems unusual that there needs to be a special case for multiple lines.

also this if length > 1 else pattern is repeated in the setMultilineText function.

please simplify

ANAMASGARD commented:
Sir @tdhock  Fixed the axis title font size issue.

Axis titles now scale correctly with `theme(text=element_text(size=30))`.

Changes:
R/z_animint.R :-  Extract axis title sizes from theme
inst/htmljs/animint.js :-  Apply those sizes to axis title text elements

Tested with different font sizes (10, 30, 50) - all working locally on my system .

Please Sir review it and give your feedback !!!

tdhock commented on inst/htmljs/animint.js:
```diff
@@ -125,24 +125,61 @@ var animint = function (to_select, json_file) {
+   * Converts <br/> tags to <tspan> elements for proper SVG rendering
+   */
+  var setMultilineText = function(textElement, text) {
+                textElement.each(function(d) {
```
is this each necessary? can it be removed?

is the setMultilineText necessary as a separate function from measureText? can they be combined?

tdhock reviewed (changes requested):
please revise and explain

tdhock commented:
please click Resolve conversation so reviewers can see which things you have already fixed.

ANAMASGARD commented:
Updated screenshot showing multiline text support working correctly:

<img width="3072" height="1840" alt="Screenshot From 2025-11-28 21-13-10" src="https://github.com/user-attachments/assets/555e61b2-f6a3-43dd-824c-d36d9262a7cb" />


 - 3-line plot title does not overlap plot area
 - X/Y axis titles with consistent multiline spacing
 - Legend title with multiline support
 - geom_text labels with multiline text

ANAMASGARD committed:
feat: implement multiline text support and fix title positioning

ANAMASGARD committed:
fix: restore axispaddingx=30 to prevent x-axis title overlap with rotated tick labels

ANAMASGARD committed:
Skip ghpages org-repo tests on PR CI; run JS_coverage under xvfb

ANAMASGARD committed:
Merge branch 'master' into multiline-text-support-221

ANAMASGARD committed:
Merge branch 'master' into multiline-text-support-221

tdhock commented on .github/workflows/tests.yaml:
```diff
@@ -48,14 +49,18 @@ jobs:
           elif [ "$TEST_SUITE" == "JS_coverage" ]; then
             echo "Running testthat with JS coverage collection..."
-            Rscript -e "source('tests/testthat.R', chdir = TRUE)"
+            xvfb-run -a Rscript -e "source('tests/testthat.R', chdir = TRUE)"
```
these changes should be moved to a different PR

ANAMASGARD committed:
Merge branch 'master' into multiline-text-support-221

ANAMASGARD committed:
Merge branch 'master' into multiline-text-support-221

ANAMASGARD committed:
Merge branch 'master' into multiline-text-support-221

ANAMASGARD committed:
revert: remove CI workflow changes from multiline PR

Per maintainer review on PR #261, move xvfb JS_coverage setup and
ghpages test skip logic out of this feature branch. Restore
tests.yaml and test-compiler-ghpages.R to match master.

ANAMASGARD committed:
Merge branch 'master' into multiline-text-support-221

tdhock commented on NEWS.md:
```diff
@@ -22,6 +22,11 @@
 
 - `update_axes`: Fixed issue #273 where axis tick text font-size was inconsistent between plots with and without `update_axes`. Previously, plots using `theme_animint(update_axes="x")` would lose `theme(axis.text = element_text(size=...))` styling after axis updates.
 
+# Changes in version 2025.11.7 (PR#261)
```
please combine news items, and remove (PR#221) which is misleading (221 is issue not PR). also please increment version in DESCRIPTION

ANAMASGARD committed:
Merge branch 'master' into multiline-text-support-221

ANAMASGARD committed:
Combine NEWS for multiline text and bump version to 2026.7.29.
Address Sir tdhock review on PR#261: single changelog entry with issue #221
(not PR#221), and increment DESCRIPTION version.

tdhock commented on tests/testthat/test-compiler-multiline-text.R:
```diff
@@ -0,0 +1,92 @@
+  )
+  info <- animint2dir(viz, "test-xaxis-multiline", open.browser = FALSE)
+  json <- RJSONIO::fromJSON(file.path(info$out.dir, "plot.json"))
+  expect_true(grepl("<br/>", json$plots$plot1$xtitle, fixed = TRUE))
```
use expect_match instead please

tdhock commented on tests/testthat/test-compiler-multiline-text.R:
```diff
@@ -0,0 +1,92 @@
+      xlab("X Axis\nLine 2")
+  )
+  info <- animint2dir(viz, "test-xaxis-multiline", open.browser = FALSE)
+  json <- RJSONIO::fromJSON(file.path(info$out.dir, "plot.json"))
```
plot.json is already in `info`, please delete this line

ANAMASGARD committed:
Merge branch 'master' into multiline-text-support-221

ANAMASGARD committed:
Use info$plots and expect_match in multiline compiler tests.

Address Sir tdhock review on PR#261: avoid re-reading plot.json and
replace expect_true(grepl(...)) with expect_match(..., fixed=TRUE).

ANAMASGARD committed:
Merge branch 'master' into multiline-text-support-221

ANAMASGARD committed:
Address Sir tdhock review: undo JS indent and tighten multiline tests.

Restore master indent on TODO cleanup. Drop redundant expect_match,
use expect_equal/expect_match, and replace the legend loop with
named title access.

tdhock commented:
great thanks
